File: src/gazebo_sim/gazebo_core/core_algorithms/core_algorithms/pid_controller.py
#!/usr/bin/env python3
import sys
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from std_msgs.msg import Float64MultiArray
from sensor_msgs.msg import JointState

class MotorTestBenchPIDNode(Node):
    def __init__(self):
        super().__init__("MotorTestBenchPIDNode")
        self.get_logger().info(f'{self.get_name()} has been started...')
        self.pub_speed_ = self.create_publisher(Float64MultiArray, '/velocity_controller/commands', 10)
        self.speed_: float = float(sys.argv[1])
        self.jointStateSub_ = self.create_subscription(JointState, '/joint_states', self.jointStateSub_callback, 10)
        self.kP: float = 10.0
        self.setPoint_ = float(sys.argv[1])
        self.error = 0.0
        self.is_position_reached = False

    def jointStateSub_callback(self, msg: JointState):
        
        joint_data = dict(zip(msg.name, msg.position))
        if "rotor_joint" in joint_data:
            self.error = joint_data["rotor_joint"]  
            self.get_logger().info(f'Current Position: {joint_data["rotor_joint"]}')

        pub_msg_ = Float64MultiArray()
        error: float = self.setPoint_ - self.error
        # pub_msg_.data = [self.kP * error]
        if self.is_position_reached:
            pub_msg_.data = [0.0]
        elif abs(error) < 0.01:
            self.is_position_reached = True
            self.get_logger().info('Target Position Reached.')
            pub_msg_.data = [0.0]
        else:
            pub_msg_.data = [self.kP * error]

        self.pub_speed_.publish(pub_msg_)
        
        if self.is_position_reached:
            rclpy.shutdown()
            return




def main(args=None) -> None:
    log = get_logger("System")
    node_instance = None

    try:
        log.info("Initializing the ROS2 Client...")
        rclpy.init(args=args)

        log.info("Starting a ROS2 Node...")
        node_instance = MotorTestBenchPIDNode()
        
        rclpy.spin(node_instance)

    except KeyboardInterrupt:
        log.warn("[CTRL+C]>>> Interrupted by the User.")

    except Exception as e:
        log.error(f"Critical Error: {e}")
    
    finally:
        if node_instance is not None:
            log.info("Destroying the ROS2 Node...")
            node_instance.destroy_node()
            node_instance = None

        if rclpy.ok():
            log.info("Manually shutting down the ROS2 Client...")
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/gazebo_sim/gazebo_core/core_algorithms/src/pid_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"

using namespace std::chrono_literals;

// Type aliases for clarity
using Float64MultiArray = std_msgs::msg::Float64MultiArray;
using JointState = sensor_msgs::msg::JointState;

class MotorTestBenchPIDNode : public rclcpp::Node {
public:
  MotorTestBenchPIDNode(double setPoint_)
      : Node("MotorTestBenchPIDNode"), setPoint_(setPoint_), error_(0.0),
        kP_(10.0), is_position_reached_(false) {
    RCLCPP_INFO(this->get_logger(), "%s has been started.", this->get_name());

    pub_speed_ = this->create_publisher<Float64MultiArray>(
        "/velocity_controller/commands", 10);

    joint_state_sub_ = this->create_subscription<JointState>(
        "/joint_states", 10,
        [this](const JointState::ConstSharedPtr &msg) -> void {
          jointStateSub_callback(msg);
        });
  }

private:
  double setPoint_;
  double error_;
  double kP_;
  bool is_position_reached_;
  rclcpp::Publisher<Float64MultiArray>::SharedPtr pub_speed_;
  rclcpp::Subscription<JointState>::SharedPtr joint_state_sub_;

  void jointStateSub_callback(const JointState::ConstSharedPtr &msg) {
    (void)msg;
    (void)setPoint_;
    // (void)error_;
    (void)kP_;
    (void)is_position_reached_;
    auto pub_msg_ = Float64MultiArray();

    for (size_t i = 0; i < msg->name.size(); i++) {
      //   std::cout << msg->name[i] << " : " << msg->position[i] << "\n";
      if (msg->name[i] == "rotor_joint") {
        RCLCPP_INFO(this->get_logger(), "Current Position: %f",
                    msg->position[i]);
        this->error_ = msg->position[i];
        break;
      }
    }

    double error = this->setPoint_ - this->error_;
    if (this->is_position_reached_) {
      pub_msg_.data = {0.0};
    } else if (abs(error) < 0.01) {
      this->is_position_reached_ = true;
      RCLCPP_INFO(this->get_logger(), "Target Position Reached");
      pub_msg_.data = {0.0};
    } else {
      pub_msg_.data = {this->kP_ * error};
    }

    this->pub_speed_->publish(pub_msg_);

    if (this->is_position_reached_) {
      rclcpp::shutdown();
      return;
    }
  }
};

int main(int argc, char *argv[]) {
  auto log = rclcpp::get_logger("System");
  MotorTestBenchPIDNode::SharedPtr node_instance = nullptr;

  try {
    RCLCPP_INFO(log, "Initializing the ROS2 Client...");
    rclcpp::init(argc, argv);

    RCLCPP_INFO(log, "Starting a ROS2 Node...");
    node_instance = std::make_shared<MotorTestBenchPIDNode>(std::stod(argv[1]));

    // Use a MultiThreadedExecutor to allow callbacks to run in parallel.
    rclcpp::spin(node_instance);

    RCLCPP_INFO(log, "Shutting down the ROS2 Node...");

    RCLCPP_WARN(log, "[CTRL+C]>>> Interrupted by the User.");
    RCLCPP_INFO(log, "Destroying the ROS2 Node...");
  } catch (const std::exception &e) {
    RCLCPP_ERROR(log, "Critical Error: %s", e.what());
  }

  if (rclcpp::ok()) {
    // Manually shut down the ROS 2 client library.
    RCLCPP_INFO(log, "Manually shutting down the ROS2 client...");
    rclcpp::shutdown();
  }

  return 0;
}